{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from os.path import join\n",
    "from matplotlib import pyplot as plt\n",
    "from collections import OrderedDict\n",
    "from collections import defaultdict\n",
    "import ast\n",
    "import gc\n",
    "import Bio.PDB\n",
    "from Bio import SeqIO\n",
    "from IPython.display import display\n",
    "from collections import Counter \n",
    "import itertools\n",
    "from IPython.html.widgets.widget_float import FloatProgress\n",
    "\n",
    "%matplotlib inline\n",
    "notebook_prefix = 'ks01'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Orienting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def find_folder_upstream(folder_name, max_iterations=50):\n",
    "    current_folder = os.getcwd()\n",
    "    counter = 0\n",
    "    while os.path.basename(current_folder) != folder_name and counter < max_iterations:\n",
    "        current_folder = os.path.dirname(current_folder)\n",
    "        counter += 1\n",
    "    if not counter < max_iterations:\n",
    "        return None\n",
    "    return current_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "root_folder_name = 'HIS3InterspeciesEpistasis'\n",
    "root_folder = find_folder_upstream(root_folder_name)\n",
    "if not root_folder:\n",
    "    print 'Did not find root folder for our github repository.\\nPlease run \"ks01\" notebook from a script inside the HIS3InterspeciesEpistasis folder!'\n",
    "\n",
    "data_folder = os.path.join(root_folder, 'Data', '')\n",
    "small_tables_folder = os.path.join(root_folder, 'Data_Small_Tables', '')\n",
    "analysis_folder = os.path.join(root_folder, 'Analysis', '')\n",
    "karen_folder = os.path.join(analysis_folder, 'Karen', '')\n",
    "figures_folder = os.path.join(karen_folder, 'figures', '')\n",
    "files_dump_folder = os.path.join(karen_folder, 'files_dump', '')\n",
    "structure_predictions_folder = os.path.join(files_dump_folder, 'structure_predictions', '')\n",
    "pymol_sessions_folder = os.path.join(files_dump_folder, 'structure_visualizations/pymol_sessions/', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/karen/anaconda/lib/python2.7/site-packages/IPython/html.py:14: ShimWarning: The `IPython.html` package has been deprecated. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.\n",
      "  \"`IPython.html.widgets` has moved to `ipywidgets`.\", ShimWarning)\n"
     ]
    }
   ],
   "source": [
    "old_dir = os.getcwd()\n",
    "os.chdir(karen_folder)\n",
    "%run 'functions_dump.py'\n",
    "os.chdir(old_dir)\n",
    "image_counter = Counter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "segment_colors_list = [('S1','#00aba9'), ('S2','#ff0097'), ('S3','#a200ff'), ('S4','#b8d000'), ('S5','#1ba1e2'), ('S6','#f09609'), \n",
    "('S7','#edc951'), ('S8','#cc2a36'), ('S9','#4f372d'), ('S10','#0c457d'), ('S11','#616161'), ('S12','#800080')]\n",
    "segment_colors = OrderedDict(segment_colors_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_segment_long_name_from_file(f):\n",
    "    return f[:-12]\n",
    "\n",
    "data_all = OrderedDict()\n",
    "files = [f for f in os.listdir(data_folder) if '.csv' in f]\n",
    "for f in sorted(files, key = lambda s: int(s.split('_')[0][1:])):\n",
    "    data_all[get_segment_long_name_from_file(f)] = pd.read_table(os.path.join(data_folder, f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "title = 'Color code for segments'\n",
    "plt.figure(figsize=(1, 5))\n",
    "ax = plt.subplot(111)\n",
    "plt.tick_params(axis=\"both\", which=\"both\", bottom=\"off\", top=\"off\",  \n",
    "    labelbottom=\"off\", left=\"off\", right=\"off\", labelleft=\"off\")\n",
    "ax.spines[\"top\"].set_visible(False)  \n",
    "ax.spines[\"right\"].set_visible(False) \n",
    "ax.spines[\"bottom\"].set_visible(False)  \n",
    "ax.spines[\"left\"].set_visible(False) \n",
    "\n",
    "counter = 1\n",
    "for segment, color in sorted(segment_colors.items(), key=lambda s: int(s[0][1:]), reverse=True):\n",
    "    plt.plot(counter * np.array([1,1]), color=color, label=segment, lw=3)\n",
    "    plt.text(1.2, counter, segment, va='center')\n",
    "    counter += 1\n",
    "plt.xlim(-0.1, 1.8)\n",
    "plt.ylim(-0.1, counter + 0.1)\n",
    "plt.tight_layout()\n",
    "# save_image(image_counter, title, figures_folder, notebook_prefix)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def remove_gaps(seq):\n",
    "    return ''.join([c for c in seq if c != '-'])\n",
    "\n",
    "def get_wt_position(position_in_alignment):\n",
    "    truncated_seq = remove_gaps(aligned_Scer[:position_in_alignment+1])\n",
    "    return len(truncated_seq)\n",
    "\n",
    "def relative_to_absolute_numbering(mutation, segment, df_with_data):\n",
    "    display(df_with_data[segment][df_with_data[segment].mut_list == mutation][['mut_list', 'mut_list_Scer']])\n",
    "    return df_with_data[segment][df_with_data[segment].mut_list == mutation].mut_list_Scer.values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Scer_Uniprot = open(os.path.join(files_dump_folder, 'HIS3_saccharomyces_cerevisiae_from_Uniprot_P06633.txt')).read().rstrip()\n",
    "alignment_file = os.path.join(small_tables_folder, 'aa_seq.txt')\n",
    "for seq_record in SeqIO.parse(alignment_file, 'fasta'):\n",
    "    if seq_record.id == 'Scer':\n",
    "        aligned_Scer = str(seq_record.seq)\n",
    "        break\n",
    "assert remove_gaps(aligned_Scer) == Scer_Uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>segment</th>\n",
       "      <th>wt1</th>\n",
       "      <th>positions_alignment</th>\n",
       "      <th>len1</th>\n",
       "      <th>len2</th>\n",
       "      <th>start1</th>\n",
       "      <th>end1</th>\n",
       "      <th>wt2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>start_Scer</th>\n",
       "      <th>end_Scer</th>\n",
       "      <th>positions_Uniprot_P06633</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S1</td>\n",
       "      <td>EALGAVRGVK</td>\n",
       "      <td>[140, 141, 142, 143, 144, 145, 146, 147, 148, ...</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>140</td>\n",
       "      <td>149</td>\n",
       "      <td>EALSRAVVDL</td>\n",
       "      <td>160</td>\n",
       "      <td>169</td>\n",
       "      <td>106</td>\n",
       "      <td>135</td>\n",
       "      <td>[106, 107, 108, 109, 110, 111, 112, 113, 114, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S2</td>\n",
       "      <td>SNRPYAVVE</td>\n",
       "      <td>[170, 171, 172, 173, 174, 175, 176, 177, 178, ...</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>170</td>\n",
       "      <td>178</td>\n",
       "      <td>LSCEMIPHF</td>\n",
       "      <td>189</td>\n",
       "      <td>197</td>\n",
       "      <td>136</td>\n",
       "      <td>163</td>\n",
       "      <td>[136, 137, 138, 139, 140, 141, 142, 143, 144, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S3</td>\n",
       "      <td>LGLQREKVGD</td>\n",
       "      <td>[179, 180, 181, 182, 183, 184, 185, 186, 187, ...</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>179</td>\n",
       "      <td>188</td>\n",
       "      <td>LESFAEA</td>\n",
       "      <td>198</td>\n",
       "      <td>204</td>\n",
       "      <td>145</td>\n",
       "      <td>170</td>\n",
       "      <td>[145, 146, 147, 148, 149, 150, 151, 152, 153, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0 segment         wt1  \\\n",
       "0               S1  EALGAVRGVK   \n",
       "1               S2   SNRPYAVVE   \n",
       "2               S3  LGLQREKVGD   \n",
       "\n",
       "Unnamed: 0                                positions_alignment len1 len2  \\\n",
       "0           [140, 141, 142, 143, 144, 145, 146, 147, 148, ...   10   10   \n",
       "1           [170, 171, 172, 173, 174, 175, 176, 177, 178, ...    9    9   \n",
       "2           [179, 180, 181, 182, 183, 184, 185, 186, 187, ...   10    7   \n",
       "\n",
       "Unnamed: 0 start1 end1         wt2 start2 end2 start_Scer end_Scer  \\\n",
       "0             140  149  EALSRAVVDL    160  169        106      135   \n",
       "1             170  178   LSCEMIPHF    189  197        136      163   \n",
       "2             179  188     LESFAEA    198  204        145      170   \n",
       "\n",
       "Unnamed: 0                           positions_Uniprot_P06633  \n",
       "0           [106, 107, 108, 109, 110, 111, 112, 113, 114, ...  \n",
       "1           [136, 137, 138, 139, 140, 141, 142, 143, 144, ...  \n",
       "2           [145, 146, 147, 148, 149, 150, 151, 152, 153, ...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positions = pd.read_table(os.path.join(small_tables_folder, 'positions.csv'))\n",
    "positions = positions.set_index('Unnamed: 0').transpose().reset_index()\n",
    "positions.rename(columns={'index' : 'segment', 'positions':'positions_alignment'}, inplace=True)\n",
    "positions.reset_index(drop=True, inplace=True)\n",
    "positions['positions_alignment'] = positions['positions_alignment'].apply(lambda s: ast.literal_eval(s))\n",
    "positions['positions_Uniprot_P06633'] = positions.positions_alignment.apply(lambda l: [get_wt_position(p) for p in l])\n",
    "segment_names = positions.segment.values # not explicitly sorted\n",
    "positions.to_csv(os.path.join(files_dump_folder, 'information_about_segments.csv'), index=False)\n",
    "positions[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x144e91390>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkkAAACqCAYAAAC5z9sxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xt8XHWd//HXZ3JPmrZpk15pS6EVWJCWooDAAi4uK1UQ\nEKGsLlh+uAoI62V1EXXbiiwXF8Ef119BioDQH3YVlB9qdcvFC0gFCqxoK7Q0bZKmaZqmaZLmNt/f\nH+c76cnkTDKTSydp38/HYx7nzDnfOd/P+X7PyXzmfE9mzDmHiIiIiPQUy3YAIiIiIiORkiQRERGR\nCEqSRERERCIoSRIRERGJoCRJREREJIKSJBEREZEISpJkWJjZDDPbbWbWR5kmMzt0/0XVNzO70sy2\n+bjLsh2P9M3M4mZ2mJ+/18y+nu2YhsNIO09EDiZKkgQAM3vXzFp8glBjZivMrHig23PObXHOjXX+\ni7jM7FkzuzypTKlz7t1Bhj4kzCwXuA34kI+7IdsxDbVwUnGA6P6SN+fclc65G4e6AjOb5dttQH8r\nzewyM/tNxPJNZvZ36WxjOM+TqPNSRPZRkiQJDviIc24ssAB4H/CN7Ia0X00BCoA/ZzuQYTQqvznW\nzHJSrdof1RO022Dqykq799FuIpImJUkSZgDOuRrg58AxAGY21cyeMrN6M9tgZld0v8Ds/Wa21swa\n/RWo//TLuz+Bm9m3gb8F7vJXqv63LxMeLhlrZg+b2Xb/KfvroTouM7PfmNl3zGynmb1jZh8Orf+0\nX7bbTy+J3DmzfDO7w8yqzGyrmd1uZnlmNhf4iy/WYGa/jnhtgZk9YmY7zKzBzP5gZhWh2B8ws2oz\n22JmNySGGf3+32ZmdT62q8NXJvwn+RvM7Hd+WOUpM5tgZo/6Nv2Dmc0MxXGkma32ffFnM/tEaN0K\nM7vLzJ72bfGimc326573/fuGX/cJM5toZj/z+1Pvy0QfGEG7VfqY1prZqaF1MTO73szeDq2f7tcd\nHYq3xsyu66sv/LrTfTt+1cxqgAf98q/4Nt5qZosJJR9+37+V9PovmVmtr+PTobIT/H4n2vcGi7ja\n4yXaZJdvtxMt8A0Lrr5uM7OHzKw0Vdv1p69+8+vD58kKC4YWV/uyzyYdH3Ezu8rMNgAb/LKTzezl\n0HH7Ab888rxMim0wx/1/WnDcv+1jCh/3Pa6kmdkSM3sk9PwkC86JBjN7zcxOD6171sy+ZWa/9XH/\nwswmhNafGnrtZjO71C/P9zFt9sfiPWZWMNB+k4OEc04PPQA2AX/n52cA/wMs9c9fAO4E8oB5wHbg\nDL/u98An/XwxcIKfnwV0ATH//Fng8qQ6u4DD/PzDwE/8NmYB64HFft1lQBtwOcEb/eeAqlCdjcAc\n/3wycFSKffyWj3eif/wOWJYUr6V47T8DTxFcbTLgOGCMX/cT4B6gECgHXgI+49d9zrflVGAc8KuI\ndtkAHAqUAn8iSNg+SPAh5gfA90P7Wglc6mOYB9QBR/r1K/zz4/1rHwUeC+1DHJgdev4fPu4YkAOc\n0sfx8Y/AeF/2i0ANkO/XfQV4PdQH7wXKgDFANfAFIB8oAd6fRl+cDnT4+PJ8m3/Y13kUUAT8kJ7H\nzwrgW0mvX+L362ygGRjn168EHvPbPcq36Qsp9rvXcUFwHG7w64qB/wIeTvH6y6K2Tc/zrb9+S97P\nRuAU3zZ3AL9J6uNfEhxrBb4fdvr+iwGL/POyVOflEB73bwHT/HGzhp7Hfff+++dLEm0ITAd2AP/g\nn5/pn08MxfxX4HAf17PAf4T6azdwke/7MuBYv+524EnfNiV+v27M9t9ePUb2I+sB6DEyHv6P1m7/\nB3QTQVJUABzi33CKQ2X/A3jQzz/v/8BNTNpeOklSHDjM//FuA44IrftnYI2fvwzYEFpX5F87ieBN\naidwPlDYzz6+nfjD65+fBWzy84eG44147WLgt8B7k5ZPAvYCBaFli4D/9vP/nXjj8M/PjGiXr4XW\n/yfw/0LPPwq86ucvAp5Pqv8+4Jt+fgWwPLTubOCt5PYOPV9G8EZ3+ACOl52JtiBI6j4aUWYR8EoG\nfbHRz5/u2zQvtP77+DdC/3wufSdJzeG+BGqBE/yx1o5P6Py6G+g/SQpv69fA50LP3+O32evYIf0k\nKa1+82XDCVQJ0AlMD5U9PbT+U8BLSXX/Hrg01Xk5hMf9P4fW/T3pJ0lfBX6QVN8vgH8KxXx9aN2V\nwDN+/jrgv1Lsyx56fkj4QOKY00OPVA8Nt0nYx5xzE5xzs51z1zjn2gg+Ce50zrWEym0m+LQHwafq\nI4C/+EvxHxlAveVALsEn+qg6ALYlZpxzrX52jI/rYoI/lDV+GOWIFPVMi6hjamKz/cT4MMEn9JV+\nuOdmC+75mEXwib7GgqHABoLEpSJU55bQdsLzCbWh+daI52P8/CzgJF9Poq5/JLh6lrAtNN8Sem2U\nW4F3gNV+SOTfUhU0s381s7f8EEYDMJag3yC48rgx4mUz/PajRPXFtNDzOudcR1L5LUnl+7pPqN45\nFw89T7RFBcEVhq2hdVF90pdpvv5wLLn07IeEToLjI1kewYePhEz6rTte51wzQcIabrvwviXHmoh3\nOukZquM+OYa+zAIuSjrOTyG4bzAhVXtFHnN+iLAYeCWxXYJbCiZmEJcchJQkSVjUm041MMHMSkLL\nZgJVAM65d5xz/+icqyB4011lZkUR2+krCdlB8IYxK7RsVqKO/jjnfuWcO4vgj+h64P4URasi6qhO\ns44u59wNzrmjgZOBcwiGvbYQfKKe6BPMMufceOfcsf6lNQRX4xJmMnBbgOd8PYm6xjrnPj+QjTnn\nmp1z/+qcOxw4F/iSmX0wuZwF9x99BbjQ11lGcNUxcbxsIRj6iIo3ajn03xfJx0sNwRtguHx/iW2U\nOoLEJdwnM1KUjYoDgjiTY++gZ3KbUElSn1vwX6OTyCxxCOuO18zGABPoea6EY64muEoa1n3+0k8b\nDvK4T+6vsGaCpCUhnABtIbiqFD7OS51z3+kr1tBr50Qs30GQTB0d2u5459y4NLYpBzElSdIn59xW\ngsvzN/mbOI8F/hfwCICZfdLMElcUGgn+6CY+wYeTrlqCobWoOuLAE8CNZjbGzGYR3PfySFT5MDOb\nZGbn+jeeDoJL6l0piq8EvmFm5T7mbybV0dd3Op1hZsf4G0/3+Lq6nHPbgNXA7WZWaoHDzOw0/9In\ngH8xs2lmNp5gKGGgngbeY2afMrNcC246f18fV86SbSPUB2b2ETNLJDFNBMlDPOJ1pQT7W+9vfv13\nvyzhAeAGM5vjt/teC75n6mlgipld6183xsxO8K/pry+SPQF82syO8n3972nucw/+WPsxsNTMiszs\nSII3/VTqCNoknOw9DnzRzA71ScqNwMqkK1cJfwD2mtm/+fOnBLgZWOucq4won46FFtyMnU8wVPii\ncy5Vsv8MMNfMFplZjpldTHAf1tN+fcrzEgZ93F9rZtP9sZB8lXIdsMgfx+8DLgytexQ4x8zOsuAG\n8EILbsafRv9+CJxpZhf6/Z1gZvOcc47gw9Mdtu/G8+lmdlYa25SDmJIkSejrE+UlwGyCT6X/RXAP\nzLN+3YeBP5nZboIbIy/2w3TJ2/we8AkL/svpjoj11xJ80ttIcKP4o865FWnEGwO+RPDJeAdwGsHQ\nW5RvA38E3iC40fiPBG9wyduMMgVYRZAI/ongvohH/bpLCW5Mfotg6ONH7PtkfD/Bm8kbwCvA/wM6\nQ2+oaV8Ncc7tIbh3ZxFBX1QTvOGm+x86S4GH/XDDhQT39fzazJoIbpy+2zkX9R9uv/SPDQT3krTQ\ncyjluwRviqvNrJEgaSry8f49wVWqbf71Z/jX9NcXyfv+C4KblNf47fx3mvvcvYnQ/DUENxPXENwY\n/xjBPXFR9bb6uH7n2+0Egv+2e4TgOH2HoD2uTfH6duAjBDfibyW4F2sKwf1lUbH1Fzs+3qVAPcGN\n1J9KVdY5t5PgvrZ/JTg//pXgqz52+iJR52XYYI77X7Kvb/8rabvfJLjis5PgfqQfhmLeCnwMuJ4g\nSd3s4068X6VsL+fcFmChL78TeA1IXN26jqD9XzKzXQTn5XtSbUsE/H9siMj+YcFXF9zrnJvdb2HZ\nL8zsZmCyc25xtmPpj5mtALY45wZ0JS1b/NXhjQQ340ddcRMZkXQlSWQY+aGCs/2l/+kEn5p/nO24\nDmZmdoSZvdfPn0AwfKw+GX7748s/RYaUkiSR4WUE/2q/k2C47U8EiZJkTynwYzPbQ3B/0Xeccz/L\nckzpGs2X/kdz7HKQ0nCbiIiISARdSRIRERGJkJtuwWXLlrklSzRKICIiIqPCoO+D05UkERERkQhK\nkkREREQiKEkSERERiaAkSURERCSCkiQRERGRCEqSRERERCIoSRIRERGJoCRJREREJIKSJBEREZEI\nSpJEREREIihJEhEREYmgJElEREQkgpIkERERkQhKkkREREQiKEkSERERiaAkSURERCSCkiQRERGR\nCEqSRERERCIoSRIRERGJoCRJREREJIKSJBEREZEISpJEREREIihJEhEREYmgJElEREQkgpIkERER\nkQhKkkREREQiKEkSERERiaAkSURERCSCkiQRERGRCEqSRERERCIoSRIRERGJoCRJREREJIKSJBER\nEZEISpJEREREIihJEhEREYmgJElEREQkgpIkERERkQhKkkREREQiKEkSERERiTCik6Qbb7yRY445\nhnnz5rFgwQLWrl3L3Xffzdy5c8nJyWHnzp3dZR977DHmzZvHvHnzOPXUU3nzzTezGHnmMtlXgGuv\nvZa5c+cyf/581q1bl6Wo94+otlmzZg3HH388xx13HKeddhobN27MdpijXlQ7A3z961/niCOO4Oij\nj+auu+7KcpSjU1TbfupTn+LII4/k2GOP5YorrqCrqyvbYY5YqY5NCP4WlpaWZjE6OaA559J6LF26\n1O1PL774ojv55JNdR0eHc865+vp6V1NT49atW+c2b97sZs+e7err63uU37Vrl3POuZ///OfuxBNP\n3K/xDkam+/rMM8+4hQsXOuece+mll0bVvmYqqm2qq6vde97zHrd+/XrnnHP33HOPW7x4cTbDHPVS\nHYMrVqxwl112WXe5urq6LEU4eqU6hn/+8593l7nkkkvcfffdl60QR7RUx6Zzzv3xj390//RP/+RK\nS0uzGaKMXGnnOKkeudlO0lKpqamhvLyc3NwgxAkTJgAwZcoUIEjuwk466aQe81VVVfsp0sHLdF+f\neuopLr30UgBOPPFEGhsbqa2tZfLkyfsx6v0jVdvEYjEaGxsBaGxsZNq0aVmL8UCQqp3vvfdeHn/8\n8e5y5eXlWYlvNEvVtlOnTu0uc8IJJ7B169asxDfSpWq/eDzOV77yFR5//HGefPLJbIYoB7ARO9x2\n1llnUVlZyZFHHsnVV1/NCy+8kPZrH3jgAc4+++xhjG5oZbqvVVVVzJgxo/v59OnTR1VSmIlUbXP/\n/fdz9tlnM3PmTB599FGuu+66LEc6uqVq53feeYeVK1fy/ve/n4985CO8/fbbWY509Onv/O7s7OSR\nRx7hwx/+cJYiHNlStd9dd93Feeedx+TJk3t9kBQZKiM2SSopKeHVV19l+fLlVFRUsGjRIh5++OF+\nX/fss8+yYsUKbrnllv0Q5dAY6L4eDKLa5gc/+AG33347v/jFL6isrGTx4sV88YtfzHaoo1qqdm5r\na6O4uJi1a9dyxRVXcPnll2c71FGnv/P7qquu4vTTT+eUU07JYpQjV1T73XTTTfzoRz/i85//fLbD\nkwNduuNy+/uepGSrVq1y5557bvfz5Pt0nHPu9ddfd3PmzHFvv/32/g5vSPW3r5/97GfdypUru58f\nccQRbtu2bfs1xmxZtWqVO/vss92cOXO6l1VWVrqjjz46i1EdeFatWuXOOeccd9RRR7l33323e/m4\nceOyGNWBIXx+L1261J1//vlZjmh0WbVqlcvLy3NTp051s2fPdoceeqiLxWJu7ty52Q5NRp5B35M0\nYq8kbdiwocel/XXr1jFr1qzu54kdSKisrOTjH/84jzzyCIcffvh+jXWwMt3Xc889t/uT6EsvvcT4\n8eMPyPuRILpt5syZQ2NjI3/9618BWL16NUcddVS2QjwgRLXzoYceynnnnceaNWsAeO655zjiiCOy\nFeKoler8fuCBB1i9enWPe76kt6j2u/LKK6murmbjxo1s2rSJ4uJiNmzYkMUo5UA1Ym/c3rNnD9dc\ncw2NjY3k5uYyZ84cli9fzp133smtt95KbW0t8+bNY+HChSxfvpwbbriBnTt3ctVVV+GcIy8vj5df\nfjnbu5GWTPd14cKFPPPMM8yZM4eSkhJWrFiR7V0YNqna5swzz+TjH/84OTk5lJWV8eCDD2Y71FEt\nVTvn5OTwyU9+kttvv53S0lIeeOCBbIc66qRq28mTJ3PooYdy0kknYWZccMEFfOMb38h2uCNOqvYL\nM7MsRScHOgtfoejLsmXL3JIlS4Y5HBEREZEhMejsecQOt4mIiIhkk5IkERERkQhKkkREREQiKEkS\nERERiaAkSURERCRCztKlS9Mq+Pzzzy8944wzMtp4R2sHDe80QAwaNzf2muaV5BHvjEeWySvJIycv\nJ+MdGkidmdTV2tbBxuqdlBTm09kVZ2P1TmJmVNbu6jUNlykpzCcvN7P9SdQVtb2+6sy0nijxvXtp\nq9xKrKgQ19lJW+VWMKO9eluvaTplwmUtd3i+eSIR80Di66vMcMbsOluJN24EixFvquw1tbwSiHdm\nXMbySrBY3rDFm05cme7DcMQLQGsnvNMIMYPNTfumJXnQGY9e11eZkjzIG4bPl4k4BxpXVJlBxNkV\nb6Wl7R0gRmv7ZnJiJTjX2WNZqmlfZXNiJcRsePo6OeZM4uqrzHDGLMNq2WA3MGzfk9TR2sHPPvMz\ndlXuYveW3ZROL6Wpqql7Om7mOEqnloLB7qrdPcqMmzmOcTPGcc7955BXlP6BOZA6M6mrta2Da25/\niuodu5lUNgYzqN6xm6q63UwpL2Xbjqbu6fRJ45g6oRQzqN25h2nlY7nzix+jqCC9/UnUtWX7Lqrq\ndvfYXl91zqgYl1E9UeJ797J5yU20b99O3sSJALTX1dFRU0vupAo6t9d1T/OmTCZ/UkWfZcJlC6ZO\nYdayrxErLBxwfH3F3LZtW48Y0omvrzLDGbPrbKX1uavp2rMF17QVGzMVt6emexorPQQrngoY8Zbq\ntMvESg8hNmYGRWfcjeUWDXm88eZqrGhSn3Flug/DES8QJB6feR4q98CWJpheAlXNMHMMTC0J/kG4\nqqXnur7KzBwDM0rh/tOhaAj/fCbirGqGyUWZxxVVZhBxdsVb+dOWz9DaXsneji0U5s+kMDfox70d\nVezt2EJB3nTaOqp6TfsqW5g/k6K8GRw9435yYkPb18kxZxJXX2WGM2YZ+YZtuK1xcyO7q3aTk5PD\n3oa9xNvjPaaxWIyGTQ3s2rirV5lYLEZTdRONlY3DXmcmdVXW7qJ6x25KivLZvK2BTTUNxGIxdu1p\npbOzq8c0J2bdZUqK8qmpb2LL9vT3J1FXYvvh7fVVZ6b1RGmv3kb79u3kFBXTtrWKtq1VmMXo3L0b\nOjp7TC0W67dMuGz79u2019QOKr6+Yk6OIZ34+ioznDHHmyqDhMNiuPZd0NXZc2o5xHe/S7xpU2Zl\nLId4czXxpi3DE29eSf9xZboPwxAvEFxRqWqGHIOGdmiPB9NYDDbtho1Nvdf1VSYWg+rmIOkajjjH\n5A0srqgyg4iztX0zezuqMMuho6sBI0ZL+yZa2zd2L3OuPXLaV1kjxt6Oava2Vw5t+0XEnElcfZUZ\nzphl5Bu2JGncrHGMnT6WeDxOYVkhOfk5PaYu7iibXcb4w8b3KuPijtJppYybOW7Y68ykrpmTxzOt\nfCwtezuYNaWM2VPLcM4xfkwRebk5PabxuOsu07K3g6kTS5kxKf39SdSV2H54e33VmWk9UfKnTSF/\n0iS6WlsoOGQ6BYdMx7k4uWPHQn5ej6lzrt8y4bL5kyaRP3Xof0IlEXNyDOnE11eZ4Yw5VjqTWMk0\nnItj+eMhJ6/n1MWJjT2UWOnszMq4OLGSacRKZwxPvB3N/ceV6T4MQ7wAzCoNrq7EHZTlQ35OMI07\nmD0WDivtva6vMnEH00qCKzXDEWdzx8DiiioziDiL8mdRmBecD3k5ZTjiFOfPpij/sO5lZvmR077K\nOuIU5k2jMH/m0LZfRMyZxNVXmeGMWUa+Yf3G7Y7WDhorGymuKKalrqXXNJGYRJUZN3NcRkNtg6kz\nk7pa2zrYsr2xOxHZsr2R8nHF7Ghs6TUNl5kxaVzGQ2CJuqK211edgxlqS4jv3Ut7TW13ctBeU0tu\n2Xg6G3b1mqZTJlx2qIetkmMeSHx9lRnOmF1nK/GmLVhROa51R69pInHItEysdMbQD12F4k0nrkz3\nYTjiBYKhrMo9UFEIdXv3TRMJRNS6vsrMHDO0Q23JcQ40rqgyg4izK97K3vZK8nIr6Ois604SwstS\nTfsqW5g/c9iGrZJjziSuvsoMZ8wyrAb9jdv6WRIRERE5EOlnSURERESGg5IkERERkQhKkkREREQi\nKEkSERERiaAkSURERCSCkiQRERGRCEqSRERERCIoSRIRERGJcMAlSTfeeCPHHHMM8+bNY8GCBaxd\nu5a7776buXPnkpOTw86dO7vLrl+/npNPPpnCwkK++93vZjHq4RPVHmvWrOH444/n2GOPZfHixcTj\n8WyHmTXJ7fPyyy9z2mmnsWDBAo477jimT5/OBRdckO0wR7VMzsmDVVQbXXHFFcyfP5/58+dz0UUX\n0dLSku0whei+Wrx4MYcddhjHHXccCxYs4I033sh2mDJEhuG79bPnpZde4plnnmHdunXk5uayc+dO\n2tvbyc/P55xzzuGMM87oUX7ixInceeedPPnkk9kJeJhFtUdbWxsnnngizz77LIcffjhLly7loYce\n4vLLL892uPtdquPlhRde6C5z4YUXct5552UxytEt03PyYJSqje644w7GjAl+fuTLX/4yd911F1/9\n6lezHO3BLdXfVIDbbruN888/P8sRylA7oJKkmpoaysvLyc0NdmvChAkATJkyBYDkn2ApLy+nvLyc\np59+ev8Gup9EtceOHTsoKCjg8MMPB+BDH/oQN91000GZJKU6XhJ2797NmjVreOihh7IQ3YEh03Py\nYNTfceico7W1FbNB/8KCDFJffXUwX5E/kB1Qw21nnXUWlZWVHHnkkVx99dU9rggcjKLao7y8nM7O\nTl599VUAVq1axdatW7McaXb0d7w89dRTfOhDH+r+NC+Z0znZv77a6PLLL2fq1KmsX7+ea665JotR\nCvTdV9dffz3z58/ny1/+Mh0dHVmMUobSAZUklZSU8Oqrr7J8+XIqKipYtGgRDz/8cLbDyppU7bFy\n5Uq+8IUvcNJJJzF27FhycnKyHWpW9He8PP7441xyySVZjHD00znZv77a6MEHH6SmpoajjjqKlStX\nZjlSSdVXN998M+vXr2ft2rXU19dzyy23ZDtUGSrOubQeS5cudaPNqlWr3Lnnntv9fPbs2a6+vr5X\nuaVLl7rbbrttf4aWFcnt4Zxzq1evdhdffHGWIhpZwu1TV1fnysvLXVtbW5ajOrCke04ezKLO0xde\neMGdc845WYpIUonqq+eee059NXKkneOkehxQV5I2bNjA22+/3f183bp1zJo1q/t5YqejpFo+mqVq\nj7q6OgDa2tq45ZZb+NznPpetELOqr+Nl1apVfPSjHyU/Pz9b4R0QBnNOHiyi2mjmzJm88847QNBG\nP/3pTznyyCOzFaJ4qY7nbdu2AUFfPfnkkxxzzDHZClGG2AF14/aePXu45ppraGxsJDc3lzlz5rB8\n+XLuvPNObr31Vmpra5k3bx4LFy5k+fLl1NbW8r73vY+mpiZisRjf+973eOuttw6Ye1BStcfNN9/M\n008/jXOOq6666qD9D6NU7QPwxBNPcN1112U5wtEv03PyYBTVRvfddx/nnXceTU1NOOeYN28e9957\nb7ZDPeilOp4/8YlPsGPHDpxzzJ8/n/vuuy/bocoQsXQ/xS1btswtWbJkmMMRERERGRKD/pfQA2q4\nTURERGSoKEkSERERiaAkSURERCSCkiQRERGRCEqSRERERCLkLF26NK2Czz///NL9+a/i7e3tbN++\nnYKCArq6uti+fTtmRn19fa9puExBQUHG3yCdqCtqe1F1DeQbqtv27mXb1i2YGXXbanpNCwqL6Ors\n7FGmoLCo+zeChrKuvuocbN3pcPE24nu3gRnxtrpeU8spANeVcRnLKcBs6GNOxJtOXJnuw3DEC9Da\n1ck7e/YQM9jc3NI9LcnNodPFI9elU6YkN4e82PB8tkoVcyaxR5Ud6pg7W6HpHSAGzZv3TXNLIN4Z\nvS6dMrklEMsbsjC74xxoXEMdT0JrZ5zNTZ0U5Rqdccfmpk4MqGru6jVNp0y4bF5seH7fLhHzQOJL\nt8xwxX6QWzbYDYzI70lqb2/noYceoqGhgXHjxgHQ0NBAQ0MD48ePZ9euXd3TCRMmMH78eAAaGxsp\nKyvj05/+dNpfApioq76+noaGhh7bS65zwoQJTJgwIaPtQ5C03HPTMupqt7GjtoYJFZPYWbe9e1o+\neQoTyicBsHNHHTtqayifPIWKyVO56mtLKCgsHLK6+qozvG4gdafDxdto3nQvnW11uPYdWN5EXEd9\n9zSWX0Esvwwwutrr0y4Ty68gp6CcktlXYrGCIY+3q72eWN74PuPKdB+GI14Iko3PvPJHKptb2NLa\nwvTCIqr2tjKzqJipRYUYRlVra4916ZSZWVTMjJJi7j/+fRTlDO2fjlQxZxJ7VNlte/cyvbhoyGLu\nbIUXPwMtldC8BYqnQ0sVlMyEwqlgBq1VPdelU6ZkJhTPgA/cD7lFg2/PRJytVVAwOfO4hjqehNbO\nOF99sYFtLV1UFMbAYFtLF9XNXUwpjrGtJd49nVaSw5SinD7LhMtOL8nl1g+UUZQ7tEl8Iuaq5s4e\nMaQTX7p4+XTgAAAOl0lEQVRlhit2GbwR2SOJhKWwsJC6ujrq6uqIxWI0NzfT2dnZY2pm3WUKCwvZ\ntWsX9fX1GdeV2H54e8l1mlnG2weo21ZDfV1wVap59266Ojt7TGOxHGqrt1JbvbW7TCyWQ33ddnbU\n1gxpXX3VGV43kLrTEW+rCxIOi+G6WjDX1XNqMbr2bifeti2zMhajq30n8bYdwxNvTlH/cWW6D8MQ\nLwRXUqpaWsmJGQ3t7bS7OA3t7cRixqbmZjY2N/dal06ZWMyobm2lsqVlv8WcSexRZcfk5Q5pzM2b\ng0SCHGhvgHh7MLUYNG+CPRt7r0unjMWgtRqaK4ckzO44c8cMLK6hjiehqrmLbS1dlOQZlXs6qWzq\nJMdgd3ucznjPaY71XyZcttYnIUMtEXNyDOnEl26Z4YpdBm9EJkkTJ06krKyMtrY2KioqqKiowDlH\nSUkJeXl5PaZAd5m2tjbGjx/PxIkTM64rsf3w9pLrBDLePkDFlKlMrJgUbG/sWHLy8npM4y7O5GmH\nMHnaId1l4i7OxIpJlE+eOqR19VVneN1A6k5HrKCCnPyJxJ3DcopxltNz6uLkFE4iVjAlszIuTk7+\nBGIF5cMTb9fe/uPKdB+GIV6AWSXFTC8uIu4cZfn55MdilOXnE3eO2SUlHFZS0mtdOmXizjGtqIiZ\nxcX7LeZMYo8q29zZOaQxl8yCoulAHPLLIJYfTF0cSmbDmMN6r0unjItD0bTgCs5QxtnZPLC4hjqe\nhOklOUwpzqGlwzFzTC4zS3OJOxibHyMvp+c0Tv9lwmUnF+cwrWTof6w7EXNyDOnEl26Z4YpdBm/E\nfuN2e3s79fX13QlJfX09paWlNDU19ZqGy0ycODHj39tK1BW1vai6BvJ7Xm1797Kjtoax48vYvauh\n1zSRkITLlE+eOqDhrv7q6qvOwdadDhdvI962A8sbi+vY3WuaSBwyLRMrKB/yoatwvOnElek+DEe8\nEAxfVba0UFFQQF1bW/c0kSxErUunzMzi4iEfausv5kxijyo71DF3tgZXWAoqoK1u3zSRUEStS6dM\nycyhHdpKxDnQuIY6noTWzjjVzV3dSUF1cxcTCmLsbIv3mqZTJlx2uIarEjEPJL50y2iobVgM+kav\nEZskiYiIiAyCfpZEREREZDgoSRIRERGJoCRJREREJIKSJBEREZEISpJEREREIihJEhEREYmgJElE\nREQkgpIkERERkQhKkkREREQiKEkSERERiaAkSURERCSCkiQRERGRCEqSRERERCIoSRIRERGJoCRJ\nREREJIKSJBEREZEISpJEREREIihJEhEREYmgJElEREQkgpIkERERkQhKkkREREQiKEkSERERiaAk\nSURERCSCkiQRERGRCLmZFF62bNlwxSEiIiIylNySJUtsMBsw59xQBSMiIiJywNBwm4iIiEgEJUki\nIiIiEdJKkszsw2b2FzPbYGb/NtxByeCY2btm9rqZvWZmL/tlZWa22szWm9kvzWxctuOUgJl938xq\nzeyN0LKU/WVmXzOzv5rZn83srOxELQkp+m+JmW01s1f948Ohdeq/EcLMDjGzNWb2JzN708yu9ct1\n/o0CEf13jV8+ZOdfv/ckmVkM2ACcCVQDa4FFzrm/DGbnZPiY2UbgeOdcQ2jZLUC9c+5Wn+iWOeeu\ny1qQ0s3MTgX2AA875471yyL7y8z+Bvgh8H7gEODXwFynmwuzJkX/LQGanHPfTSp7FPAY6r8Rwcym\nAFOcc+vMbAzwCvAxYDE6/0a8PvrvYobo/EvnStIJwF+dc5udcx3ASh+EjFxG7779GPADP/8D4Lz9\nGpGk5Jz7LdCQtDhVf50LrHTOdTrn3gX+SnCOSpak6D8IzsNkH0P9N2I457Y559b5+T3AnwnePHX+\njQIp+m+6Xz0k5186SdJ0YEvo+dZQEDIyOeBXZrbWzK7wyyY752ohOLCASVmLTtIxKUV/JZ+PVeh8\nHKk+b2brzOyB0HCN+m+EMrNDgfnAS6T+e6n+G6FC/fcHv2hIzj/duH1gOsU5twBYCFxtZn9LkDiF\n6fLw6KL+Gl3uAQ5zzs0HtgG3ZTke6YMfqlkF/Iu/IqG/l6NIRP8N2fmXTpJUBcwMPT/EL5MRyjlX\n46d1wJMElxNrzWwydI/jbs9ehJKGVP1VBcwIldP5OAI55+pC9zncz75L+uq/EcbMcgneYB9xzj3l\nF+v8GyWi+m8oz790kqS1wBwzm2Vm+cAi4Kfp74LsT2ZW7LNqzKwEOAt4k6DPPu2LXQY8FbkByRaj\n5xh6qv76KbDIzPLNbDYwB3h5fwUpKfXoP//GmnAB8D9+Xv038jwIvOWc+15omc6/0aNX/w3l+dfv\nz5I457rM7PPAaoKk6vvOuT9ntg+yH00GfmJmjqB/f+icW21mfwSeMLPLgc3ARdkMUvYxs8eAM4CJ\nZlYJLAFuBn6U3F/OubfM7AngLaADuEr/WZNdKfrvg2Y2H4gD7wKfBfXfSGNmpwCfBN40s9cIhtWu\nB24h4u+l+m9k6aP//nGozj/9LImIiIhIBN24LSIiIhJBSZKIiIhIBCVJIiIiIhGUJImIiIhEUJIk\nIiIiEkFJkoiIiEgEJUlyUDKzLjN71czeNLP/a2aFA9jGcjM70s9/LWndb4cq1pHEzFaY2QV+/v7E\n/g9ym+PM7MoMXzPLzN5MWrbEzL7Uz+uON7M7BhJnxLYyjltERhclSXKwanbOLXDOvZfgS8U+l+kG\nnHP/7Jz7i396fdK6U4cgxqwys5y+1jvnPhPa/8EoA64awOsy/pI359wrzrkvpFu+nzYYaNwiMkoo\nSRKB3xB8PT1m9iV/dekNM/sXv6zYzJ42s9f88k/45c+a2QIzuwko8lemHvHrmhIbN7Pv+G2+bmYX\n+WWn+9f/yMz+nHidX3ezmf2P/wXrW5ODNbP3m9nvzewVM/utmc31y2OhutaZ2dWh8r/zy14ysxIz\nKzCzB/3+vGJmZ/iyl5nZU2b238Cv/bK7fIyr2fdr6N37n9hfM/u2r+P3Zlbhlx9mZi/6fb8h3C4h\nNwGH+fa7JVWbZcLHdrOZ/cHM/uK/mTfR7j/z80vM7GEf73ozuyJU5gUzewr4k18WPi6uTRV3qP5U\nx8wCM3vOzNaa2c9t3++DHe/b7jUzuzVxlcz3x52h7f7MzE7z83/vY/+jBVdDi/3yTWa21Pfr62b2\nHr+8JNTn68zs/L62IyKAc04PPQ66B9Dkp7kEPwL8WWAB8DpQCJQQ/N7PPILf/vk/odeW+umzwAI/\nvztp+7v99OPAL/38JIKfOJgMnA40AFMJfvPr98DJwATgL6HtjI2IfQwQ8/NnAqv8/JXAE+z7Jv3x\nQB7wTijOMUAO8CXgAb/sCB9XPsHvVFUC4/y680PxT/UxXxCx/3FgoZ+/Bbjez/8MuMjPfza5nfzy\nWcAboecXRLVZX6/xy5YAXwrF9h0/fzbwKz9/OvDTUPnX/H5P9Ps9xZdpAmb6cqmOi14xJO1Dj2OG\n4Fj7HTDRL7uI4Gee8Ns/xc/fmtiu74//HdrOz4DTfLzPA0V++VeBb/j5TQQ/t5A4Jpb7+ZuB74a2\nNS7Fdr6Z7fNTDz1GykNXkuRgVWRmrxL8uOG7wPeBU4GfOOf2OueagR8Df0vwA8F/b2Y3mdmpzrmo\nqyGpnAI8DuCc2w48B7zfr3vZOVfjnHPAOuBQoBFoNbMH/Cf91ohtjgdW+asNtwN/45efSfDG7Hx9\nuwgSoGrn3Kt+2R7nXJff10f9svW+Dd7jt/Mr51yjnz8tFH8NsCbFfrY5557x86/4fQH4AMEvdAM8\nluK1yU4ldZslpBpqCy//cSieWSnKP+Wca3fO1RPsW+LXwl92zlWG4ok6LvoSdcwcARwD/MqC35n6\nOjDNzMYRJKW/8699JHqTPZxE0O+/89u6FJgZWv8TPw33xYeAuxMFfB/3tx2Rg1q/P3ArcoBqcc4t\nCC8ws8iCzrm/+mGlhcC3zezXzrlvJxWLfnFv4XJtofkuINcFPyh9AkHC8wng834+7AZgjXPuAjOb\nRXDVJN060ynTnEb5ZB2h+S72/W0JJy3ptlGyqNfVE1x1C5sAbAw9T7RvOJ5kyfElng+kDfZttOcx\nc4MfvnwS+B/n3Cnhsj5JSqWTnrdFJP7BwIDVzrlPpnhdOvueznZEDmq6kiQHq6g33t8A55lZoZmV\nEAw1/cbMpgKtzrnHgO8QDL8kazez8JtRYvu/AS729wtVEFyBeDllUMH9IOOdc78gGBI7NqLYWKDK\nzy8OLf8V8FnzNxubWRmwHphiZsf7ZWP8+t8Q/Ho2/p6VGb5sshdC8U8FPpgq9BTLXwIu9POLUpRp\nIhiOSui3zfwVnWoz+6DfhwnAPwCp/qswVXwfM7N8M5tIMMy2NqJM5HEREfe+ynoeM/9JcMysByrM\n7CRfJtfM/sZf0dllZif7l4cTlneB+RaYwb4rXS8Bp5jZ4X5bxebvTevDr4CrQzGOH+B2RA4aSpLk\nYNVruMY59xrwEMEb5YsE93K8DrwXeNkPR/w7wZWc5G0sB96wfTdgJ4a8fgK8QXDPya+Br/ghpFTx\njAWeNrPXCRKUL0aU/Q5ws5m9Qs9z+AFgi4/jNeAS51wHcDFwl5mtA1YDBcA9QI6ZvUEwtHWZL5vc\nJj8B3ia4gfkhgnunkmNOng/7IvAlX/fhBMOJyXXsJBjuecPMbvF1vkn/bXYp8E2/r78GljrnNqWI\nJ1V8bxAM5/0e+JZzbltEfJHHRXLcSS9LPma+7dv3QuAW3x6vEQxHAlwO3OOHgMN1/44gUfoTcAfB\n8BnOuR3Ap4HH/bHye4LhvL729dvABH8D+mvAGf1sR+Sgl7jBU0RkyJlZkXOu1c9fDCxyzp2f5bCA\n4L/bCG7g/262YwnzQ6hPu+DrKUQki3RPkogMp+PN7C6C4a4Ggism0j99ehUZAXQlSURERCSC7kkS\nERERiaAkSURERCSCkiQRERGRCEqSRERERCIoSRIRERGJoCRJREREJML/BxdO6CFGFG67AAAAAElF\nTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x14840d490>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "title = 'Positions of segments according to Uniprot sequence'\n",
    "plot_better(grid='', height=2)\n",
    "old_y = 2\n",
    "for row in positions.iterrows():\n",
    "    for position in row[1].positions_Uniprot_P06633:\n",
    "        new_y = np.random.choice([1,2])\n",
    "        while new_y == old_y:\n",
    "            new_y = np.random.choice([1,2])\n",
    "    x = row[1].positions_Uniprot_P06633\n",
    "    plt.plot(x, [new_y for e in x], '.', lw=3, alpha=0.7, label=row[1].segment, color=segment_colors[row[1].segment])\n",
    "    plt.text(np.median(x), new_y + 0.3, row[1].segment)\n",
    "    old_y = new_y\n",
    "plt.ylim(0,4)\n",
    "plt.yticks([])\n",
    "plt.xlabel('Positions according to Uniprot sequence')\n",
    "plt.title(title)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Keeping only the mutants carrying substitutions (no indels, no mutations in the unmutated \"central region\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_mutations(muts):\n",
    "    if not muts > 0:\n",
    "        return 0\n",
    "    else:\n",
    "        return muts.count(':') + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data = OrderedDict()\n",
    "for k,v in data_all.items():\n",
    "    data[k] = v[(v.middle == 1) & (v.nogap == 1)].copy()\n",
    "    data[k]['mut_number'] = data[k].mut_list_Scer.apply(count_mutations)\n",
    "\n",
    "data_natural = OrderedDict()\n",
    "for k,v in data_all.items():\n",
    "    data_natural[k] = v[(v.middle == 1) & (v.nogap == 1) & (v.nat_lib == 1)].copy()\n",
    "    data_natural[k]['mut_number'] = data_natural[k].mut_list_Scer.apply(count_mutations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Killing data_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "144"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all = None\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adding from_aa to notation of mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_full_mutation(mutation):\n",
    "    if mutation == '' or not mutation > 0:\n",
    "        return ''\n",
    "    position = int(mutation[:-1])\n",
    "    assert Scer_Uniprot[position-1] != mutation[-1]\n",
    "    return Scer_Uniprot[position-1] + mutation\n",
    "\n",
    "def convert_to_full_mutations(mut_combination):\n",
    "    if mut_combination == '' or not mut_combination > 0:\n",
    "        return ''\n",
    "    return ':'.join(get_full_mutation(mutation) for mutation in mut_combination.split(':'))\n",
    "    \n",
    "for segment in data:\n",
    "    data[segment]['mut_list_Scer_full'] = data[segment]['mut_list_Scer'].apply(convert_to_full_mutations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino acid properties table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# amino acid codes\n",
    "aa_long = ['alanine', 'cysteine', 'aspartic acid', 'glutamic acid', 'phenylalanine', 'glycine', \n",
    "           'histidine', 'isoleucine', 'lysine', 'leucine', 'methionine', 'asparagine', 'proline',\n",
    "           'glutamine', 'arginine', 'serine', 'threonine', 'valine', 'tryptophan', 'tyrosine', 'unknown_aa'] \n",
    "aa1 = list(\"ACDEFGHIKLMNPQRSTVWYX\")\n",
    "aa3 = \"ALA CYS ASP GLU PHE GLY HIS ILE LYS LEU MET ASN PRO GLN ARG SER THR VAL TRP TYR XXX\".split()\n",
    "aa123 = dict(zip(aa1, aa3))\n",
    "aa321 = dict(zip(aa3, aa1))\n",
    "aa_full_to_1 = dict(zip(aa_long, aa1))\n",
    "aa_1_to_full = dict(zip(aa1, aa_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Original table http://www.proteinsandproteomics.org/content/free/tables_1/table08.pdf\n",
    "\n",
    "aa_properties = pd.read_table(os.path.join(files_dump_folder, 'properties_of_amino_acids.csv'), sep=';')\n",
    "\n",
    "aa_properties['Fraction of buried among this aa'] = aa_properties['Percent buried residues'].apply(lambda s: float(s.split()[0]))\n",
    "aa_properties['Fraction of buried among all buried'] = aa_properties['Percent buried residues'].apply(lambda s: float(s.split()[1][1:-1]))\n",
    "\n",
    "aa_properties['Polarity average ranking'] = aa_properties['Ranking of amino acid polarities'].apply(lambda s: float(s.split()[0]))\n",
    "aa_properties['Polarity ranking (Radzicka and Wolfenden 1988)'] = aa_properties['Ranking of amino acid polarities'].apply(lambda s: float(s.split()[1][1:-1]))\n",
    "\n",
    "aa_properties['amino_acid'] = aa_properties['Amino acid residue'].apply(lambda s: s.lower())\n",
    "aa_properties.set_index('amino_acid', inplace=True)\n",
    "\n",
    "aa_properties.drop(['Unnamed: 8', 'Unnamed: 10', 'Percent buried residues', 'Ranking of amino acid polarities', 'Amino acid residue'], inplace=True, axis=1)\n",
    "aa_properties = aa_properties.astype(np.float, raise_on_error=False)\n",
    "\n",
    "quantitative_properties = [p for p in aa_properties.columns if 'pKa' not in p]\n",
    "\n",
    "def get_aa_properties_by_single_letter_code(aa_1):\n",
    "    return aa_properties.loc[aa_1_to_full[aa_1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Describing mutations by properties of aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def aa_property_change(mutation, aa_property):\n",
    "    assert aa_property in aa_properties.columns.values\n",
    "    if mutation == '':\n",
    "        return None\n",
    "    aa_from = mutation[0]\n",
    "    position = int(mutation[1:-1])\n",
    "    aa_to = mutation[-1]\n",
    "    properties_from = get_aa_properties_by_single_letter_code(aa_from)\n",
    "    properties_to = get_aa_properties_by_single_letter_code(aa_to)\n",
    "#     print mutation, properties_to[aa_property], properties_from[aa_property], properties_to[aa_property] - properties_from[aa_property]\n",
    "    return properties_to[aa_property] - properties_from[aa_property]\n",
    "\n",
    "\n",
    "precalculated_changes = {}\n",
    "for aa_property in quantitative_properties:\n",
    "    precalculated_changes[aa_property] = {}\n",
    "    for substitution in list(itertools.permutations([aa for aa in aa1 if aa != 'X'], 2)):\n",
    "        precalculated_changes[aa_property][substitution] = aa_property_change(substitution[0] + '123' + substitution[1], aa_property)        \n",
    "        \n",
    "def cumulative_property_change(mut_combination, aa_property, absolute_values=True, use_precalculated=True):\n",
    "    assert aa_property in aa_properties.columns.values\n",
    "    if mut_combination == '':\n",
    "        return None\n",
    "    changes = []\n",
    "    if use_precalculated:\n",
    "        for m in mut_combination.split(':'):\n",
    "            changes.append(precalculated_changes[aa_property][m[0], m[-1]])\n",
    "    else:\n",
    "        for m in mut_combination.split(':'):\n",
    "            changes.append(aa_property_change(m, aa_property))\n",
    "    if absolute_values:\n",
    "        return sum([np.abs(v) for v in changes])\n",
    "    else:\n",
    "        return sum(changes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating changes in properties:\n",
      "   Average residue mass (daltons)\n",
      "   Monoisotopic mass (daltons)\n",
      "   Occurrence in proteins, %\n",
      "   Average volume of buried residues, A3\n",
      "   Van der Waals volume, A3\n",
      "   Accessible surface area, A2\n",
      "   Fraction of buried among this aa\n",
      "   Fraction of buried among all buried\n",
      "   Polarity average ranking\n",
      "   Polarity ranking (Radzicka and Wolfenden 1988)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-9b4b8cfd7304>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0maa_property\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mquantitative_properties\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         data[segment]['Abs change in ' + aa_property.lower()] = data[segment]['mut_list_Scer_full'].apply(\n\u001b[0;32m---> 11\u001b[0;31m             lambda mut_comb: cumulative_property_change(mut_comb, aa_property))\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msegment\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_hdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiles_dump_folder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'processed_data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Data_dict_%s.hdf'\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0msegment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'data'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/karen/anaconda/lib/python2.7/site-packages/pandas/core/series.pyc\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   2218\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2219\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masobject\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2220\u001b[0;31m             \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2222\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/src/inference.pyx\u001b[0m in \u001b[0;36mpandas.lib.map_infer (pandas/lib.c:62658)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-36-9b4b8cfd7304>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(mut_comb)\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0maa_property\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mquantitative_properties\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         data[segment]['Abs change in ' + aa_property.lower()] = data[segment]['mut_list_Scer_full'].apply(\n\u001b[0;32m---> 11\u001b[0;31m             lambda mut_comb: cumulative_property_change(mut_comb, aa_property))\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msegment\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_hdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiles_dump_folder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'processed_data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Data_dict_%s.hdf'\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0msegment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'data'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-23-bf743a34958f>\u001b[0m in \u001b[0;36mcumulative_property_change\u001b[0;34m(mut_combination, aa_property, absolute_values, use_precalculated)\u001b[0m\n\u001b[1;32m     25\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0muse_precalculated\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mm\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmut_combination\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m':'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m             \u001b[0mchanges\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprecalculated_changes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maa_property\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mm\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmut_combination\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m':'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "print 'Calculating changes in properties:'\n",
    "for p in quantitative_properties:\n",
    "    print '  ', p\n",
    "sys.stdout.flush()\n",
    "\n",
    "f = FloatProgress(min=0, max=len(data.keys()))\n",
    "display(f)\n",
    "for segment in data:\n",
    "    for aa_property in quantitative_properties:\n",
    "        data[segment]['Abs change in ' + aa_property.lower()] = data[segment]['mut_list_Scer_full'].apply(\n",
    "            lambda mut_comb: cumulative_property_change(mut_comb, aa_property))\n",
    "    data[segment].to_hdf(os.path.join(files_dump_folder, 'processed_data', 'Data_dict_%s.hdf' %segment), 'data')\n",
    "    f.value += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Unique single mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "unique_single_mutations = OrderedDict()\n",
    "for segment_name, segment_data in data.items():\n",
    "    without_wt = segment_data.dropna(subset=['mut_list_Scer'])\n",
    "    unique_single_mutations[segment_name] = set(':'.join(without_wt.mut_list_Scer.values).split(':'))\n",
    "unique_single_mutations['all_segments'] = set().union(*unique_single_mutations.values())\n",
    "print len(unique_single_mutations['all_segments']), 'unique mutations across all segments'\n",
    "\n",
    "natural_unique_single_mutations = OrderedDict()\n",
    "for segment_name, segment_data in data_natural.items():\n",
    "    without_wt = segment_data.dropna(subset=['mut_list_Scer'])\n",
    "    natural_unique_single_mutations[segment_name] = set(':'.join(without_wt.mut_list_Scer.values).split(':'))\n",
    "natural_unique_single_mutations['all_segments'] = set().union(*natural_unique_single_mutations.values())\n",
    "print len(natural_unique_single_mutations['all_segments']), 'natural unique mutations across all segments'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Singles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data_singles = OrderedDict()\n",
    "known_single_mutations = []\n",
    "for segment_name, segment_data in data.items():\n",
    "    data_singles[segment_name] = segment_data[segment_data.mut_list.str.count(':') == 0].copy()\n",
    "    data_singles[segment_name]['position_Scer'] = data_singles[segment_name]['mut_list_Scer'].apply(lambda s: int(s[:-1]))\n",
    "    known_single_mutations += set(':'.join(data_singles[segment_name].mut_list_Scer.values).split(':'))\n",
    "positions_with_known_mutations = set([int(m[:-1]) for m in known_single_mutations])\n",
    "print 'There are %s single mutations with known effects. They are found in %s positions' % (len(known_single_mutations),\n",
    "                                                                                            len(positions_with_known_mutations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A dictionary containing fitness for every known single mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "single_mut_fitness_dict = {}\n",
    "for segment_data in data_singles.values():\n",
    "    single_mut_fitness_dict.update(dict(segment_data[['mut_list_Scer', 's']].values))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotypes consisting of mutations with known effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "wt_fitness = np.median([table[table.mut_number == 0].s.values[0] for table in data.values()])\n",
    "min_fitness = np.min([table.s.min() for table in data.values()])\n",
    "max_fitness = np.max([table.s.max() for table in data.values()])\n",
    "\n",
    "def mut_effect(fitness, wt_fitness=wt_fitness):\n",
    "    return fitness - wt_fitness\n",
    "\n",
    "def epistasis(single_mut_fitnesses, fitness_of_combination, wt_fitness=wt_fitness):\n",
    "    min_mut_effect = mut_effect(min_fitness, wt_fitness=wt_fitness) # has to be inside function for calculating foursome epistasis\n",
    "    max_mut_effect = mut_effect(max_fitness, wt_fitness=wt_fitness) # has to be inside function for calculating foursome epistasis\n",
    "    expected_mut_effect = sum([mut_effect(fitness, wt_fitness=wt_fitness) for fitness in single_mut_fitnesses])\n",
    "    expected_mut_effect = min(expected_mut_effect, max_mut_effect)\n",
    "    expected_mut_effect = max(expected_mut_effect, min_mut_effect)\n",
    "    return mut_effect(fitness_of_combination, wt_fitness=wt_fitness) - expected_mut_effect\n",
    "\n",
    "def epistasis_for_df(df):\n",
    "    mutations_combination = df['mut_list_Scer']\n",
    "    fitness_of_combination = df['s']\n",
    "    single_mut_fitnesses = []\n",
    "    for single_mutation in mutations_combination.split(':'):\n",
    "        single_mut_fitnesses.append(single_mut_fitness_dict[single_mutation])\n",
    "    return epistasis(single_mut_fitnesses, fitness_of_combination)\n",
    "\n",
    "def foursome_epistasis(panel, wt_mut_combination):\n",
    "    panel_slice = panel.major_xs(wt_mut_combination).loc['s']\n",
    "    wt_fitness = panel_slice['wild_type']\n",
    "    mutA_fitness = panel_slice['mutA']\n",
    "    mutB_fitness = panel_slice['mutB']\n",
    "    mutAB_fitness = panel_slice['mutAB']\n",
    "    return epistasis([mutA_fitness, mutB_fitness], mutAB_fitness, wt_fitness=wt_fitness)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "data_known_mutations_only = OrderedDict()\n",
    "for segment_name, segment_data in data.items():\n",
    "    without_wt = segment_data.dropna(subset=['mut_list_Scer'])\n",
    "    data_known_mutations_only[segment_name] = without_wt[without_wt.mut_list_Scer.apply(\n",
    "            lambda muts: consists_of_known_mutations(muts, data_singles[segment_name].mut_list_Scer.values))].copy()\n",
    "    data_known_mutations_only[segment_name]['epistasis'] = data_known_mutations_only[segment_name].apply(epistasis_for_df, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print '''\n",
    "\\nNow the following variables are available:\n",
    "\n",
    "- data: all data as a (sorted) dictionary in the form: 'segment_name': 'segment_data_as_pandas_dataframe'\n",
    "\n",
    "- unique_single_mutations: a dictionary with all single mutations for every segment\n",
    "\n",
    "- data_singles: a sorted dictionary in the form: 'segment_name': 'segment_data_as_pandas_dataframe' for genotypes containing a single substitution\n",
    "\n",
    "- single_mut_fitness_dict: a dictionary in the form: 'single mutation' : 'fitness'\n",
    "\n",
    "- data_known_mutations_only - a sorted dictionary in the form: 'segment_name': 'segment_data_as_pandas_dataframe', \n",
    "but only for those genotypes that consist of mutations with known effects (single mutant is measured).\n",
    "For this dictionary, epistasis is measured for every genotype.\n",
    "\n",
    "# # # # # # # # # # # # # # # # #\n",
    "\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
